File: geode-client-native/quickstart/cpp/PdxSerializer.cpp
/*
* The PdxSerializer QuickStart Example.
* This example takes the following steps:
*
* This example shows PdxSerializer usage. 
*
* 1. Create a GemFire Cache.
* 2. Get the Person from the Cache.
* 3. Populate some query Person objects on the Region.
* 4. Get the pool, get the Query Service from Pool. Pool is define in clientPdxSerializer.xml. 
* 5. Execute a query that returns a Result Set.
* 6. Execute a query that returns a Struct Set.
* 7. Execute the region shortcut/convenience query methods.
* 8. Close the Cache.
*
*/


// Include the GemFire library.
#include <gfcpp/GemfireCppCache.hpp>

// Use the "gemfire" namespace.
using namespace gemfire;

static const char * CLASSNAME = "com.example.Person";

//This demonstrates Person domain class
class Person
{
private:
  char* m_name;    
  int m_id;
  int m_age;

public:
  Person() { }

  Person(char* name, int id, int age)
  {
    m_name = name;
    m_id = id;
    m_age = age;
  }

  char* getName() const
  {
    return m_name;
  }
  int getID()
  {
    return m_id;
  }
  int getAge()
  {
    return m_age;
  }
  void setName(char* name)
  {
    m_name = name;
  }
  void setID(int id)
  {
    m_id = id;
  }
  void setAge(int age)
  {
    m_age = age;
  }
};

//This demonstrates, how to extend PdxSerializer without modifying Person domain object, necessary for serialization/desrialization to PDX format.
class PersonPdxSerializer : public PdxSerializer
{
public:

  static void deallocate(void * testObject, const char * className)
  {    
    LOGINFO("PersonPdxSerializer::deallocate called");
    if (strcmp(className, CLASSNAME) == 0) {
      Person* per = reinterpret_cast<Person*>(testObject);
      delete per;
    }
  }

  static uint32_t objectSize(void * testObject, const char * className)
  {    
    LOGINFO("PersonPdxSerializer::objectSize called");
    if (strcmp(className, CLASSNAME) == 0) {
      Person* per = reinterpret_cast<Person*>(testObject);
      uint32_t size = 0;
      size += sizeof(Person);
      if (per->getName() != NULL) {
        size += strlen(per->getName());
      }
      return size;
    }
    return 0;
  }

  UserDeallocator getDeallocator(const char * className)
  {    
    if (strcmp(className, CLASSNAME) == 0) {
      return deallocate;
    }
    return NULL;
  }

  UserObjectSizer getObjectSizer(const char * className)
  {   
    if (strcmp(className, CLASSNAME) == 0) {
      return objectSize;
    }
    return NULL;
  }

  void * fromData(const char * className, PdxReaderPtr pr)
  { 
    if (strcmp(className, CLASSNAME) == 0) {
      Person* per = new Person();

      try
      {
        per->setID(pr->readInt("m_id"));
        per->setAge(pr->readInt("m_age"));
        per->setName(pr->readString("m_name"));
      }
      catch (...)
      {
        return NULL;
      }
      return (void*) per;
    }
    return NULL;    
  }  

  bool toData(void * testObject, const char * className, PdxWriterPtr pw)
  {
    if (strcmp(className, CLASSNAME) == 0) {

      Person* per = reinterpret_cast<Person*>(testObject);

      try
      {
        pw->writeInt("m_id" ,per->getID());
        pw->writeInt("m_age" ,per->getAge());
        pw->writeString("m_name" ,per->getName());
      }
      catch (...)
      {
        return false;
      }
      return true;
    }    
    return false;
  }
};

// This PdxSerializer QuickStart example demonstrartes query on .NET objects without having corresponding java classes at server.
int main(int argc, char ** argv)
{
  try
  {
    // Create a GemFire Cache.
    CacheFactoryPtr cacheFactory = CacheFactory::createCacheFactory();

    // Create a GemFire Cache with the "clientPdxSerializer.xml" Cache XML file.
    CachePtr cachePtr = cacheFactory->set("cache-xml-file", "XMLs/clientPdxSerializer.xml")->create();          

    LOGINFO("Created the GemFire Cache");

    // Get the example Region from the Cache which is declared in the Cache XML file.
    RegionPtr regionPtr = cachePtr->getRegion("Person");       

    LOGINFO("Obtained the Region from the Cache.");

    // Register PersonPdxSerializer to serialize the domain types(Person class) as pdx format
    Serializable::registerPdxSerializer(PdxSerializerPtr(new PersonPdxSerializer));    
    LOGINFO("Registered Person Query Objects");

    // Populate the Region with some Person objects.
    Person* p1 = new Person("John", 1 /*ID*/, 23 /*age*/);
    PdxWrapperPtr pdxobj1(new PdxWrapper(p1, CLASSNAME));
    regionPtr->put("Key1", pdxobj1);

    Person* p2 = new Person("Jack", 2 /*ID*/, 20 /*age*/);
    PdxWrapperPtr pdxobj2(new PdxWrapper(p2, CLASSNAME));
    regionPtr->put("Key2", pdxobj2);

    Person* p3 = new Person("Tony", 3 /*ID*/, 35 /*age*/);
    PdxWrapperPtr pdxobj3(new PdxWrapper(p3, CLASSNAME));
    regionPtr->put("Key3", pdxobj3);

    LOGINFO("Populated some Person Objects through PdxWrapper");

    //find the pool
    PoolPtr poolPtr = PoolManager::find("examplePool");

    // Get the QueryService from the Pool.
    QueryServicePtr qrySvcPtr = poolPtr->getQueryService();

    LOGINFO("Got the QueryService from the Pool");

    // Execute a Query which returns a ResultSet.    
    QueryPtr qryPtr = qrySvcPtr->newQuery("SELECT DISTINCT * FROM /Person");
    SelectResultsPtr resultsPtr = qryPtr->execute();

    LOGINFO("ResultSet Query returned %d rows", resultsPtr->size());

    // Execute a Query which returns a StructSet.
    qryPtr = qrySvcPtr->newQuery("SELECT m_name, m_age FROM /Person WHERE m_id = 1");
    resultsPtr = qryPtr->execute();

    LOGINFO("StructSet Query returned %d rows", resultsPtr->size());

    // Iterate through the rows of the query result.
    int rowCount = 0;
    SelectResultsIterator iter = resultsPtr->getIterator();
    while (iter.hasNext())
    {
      rowCount++;
      Struct * psi = dynamic_cast<Struct*>( iter.next().ptr() );
      LOGINFO("Row %d Column 1 is named %s, value is %S", rowCount, psi->getFieldName(0), (*psi)[0]->toString()->asWChar());
      LOGINFO("Row %d Column 2 is named %s, value is %s", rowCount, psi->getFieldName(1), (*psi)[1]->toString()->asChar());
    }

    // Execute a Region Shortcut Query (convenience method).
    resultsPtr = regionPtr->query("m_id = 2");

    LOGINFO("Region Query returned %d rows", resultsPtr->size()); 

    // Execute the Region selectValue() API.
    PdxWrapperPtr pdxWrapperPtr = regionPtr->selectValue("m_id = 3");
    Person* per = reinterpret_cast<Person*>(pdxWrapperPtr->getObject());

    LOGINFO("Region selectValue() returned an item:\n Person Name = %s and Person Age = %d", per->getName(), per->getAge());

    // Execute the Region existsValue() API.
    bool existsValue = regionPtr->existsValue("m_id = 4");

    LOGINFO("Region existsValue() returned %s", existsValue ? "true" : "false");

    // Close the GemFire Cache.
    cachePtr->close();

    LOGINFO("Closed the GemFire Cache");

  }
  // An exception should not occur
  catch(const Exception & gemfireExcp)
  {    
    LOGERROR("PdxSerializer GemFire Exception: %s", gemfireExcp.getMessage());
  }
}

File: geode-client-native/quickstart/csharp/PdxSerializer.cs
/*
 * The PdxSerializer QuickStart Example.
 * This example takes the following steps:
 *
 * This example shows IPdxSerializer usage. 
 * We have used inbuilt ReflectionBasedAutoSerializer as IPdxSerializer. ReflectionBasedAutoSerializer class implements the IPdxSerializer interface
 * This auto serializer uses reflection to serialize and de-serialize class members.
 * User can use PdxIdentityField attribute on member fields to define field as identity field.
 * ReflectionBasedAutoSerializer will use this attribute to set field as identity field for Pdx data.
 *
 * AutoSerializerEx class extends the ReflectionBasedAutoSerializer class and override WriteTransform and ReadTransform methods.
 * In this method it handles serialization of .NET decimal type and Guid type.
 * 
 * PdxTypeMapper class demonstrates that how app can map .NET types to pdx types(java types)
 * 
 * Domain class should have default constructor(zero-arg).
 *
 * After that test demonstrartes query on .NET objects without having corresponding java classes at server.
 *
 * 1. Create a GemFire Cache.
 * 2. Get the Person from the Cache.
 * 3. Populate some query Person objects on the Region.
 * 4. Get the pool, get the Query Service from Pool. Pool is define in clientPdxRemoteQuery.xml. 
 * 5. Execute a query that returns a Result Set.
 * 6. Execute a query that returns a Struct Set.
 * 7. Execute the region shortcut/convenience query methods.
 * 8. Close the Cache.
 *
 */
// Use standard namespaces
using System;
using System.Reflection;
// Use the GemFire namespace
using GemStone.GemFire.Cache.Generic;


namespace GemStone.GemFire.Cache.Generic.QuickStart
{
  public class Person
  {
    private string name;
    //this is the only field used on server to create hashcode and use in equals method
    [PdxIdentityField]
    private int id;
    private int age;
    //private decimal salary;
    private Guid guid = Guid.NewGuid();

    public Person() { }

    public Person(string name, int id, int age)
    {
      this.name = name;
      this.id = id;
      this.age = age;
      //this.salary = 217773.12321M;
    }
    #region Public Properties
    public string Name
    {
      get { return name; }
    }
    public int ID
    {
      get { return id; }
    }
    public int Age
    {
      get { return age; }
    }
    #endregion
  }
  
  //This demonstrates, how to extend ReflectionBasedAutoSerializer
  public class AutoSerializerEx : ReflectionBasedAutoSerializer
  {
    public override object WriteTransform(FieldInfo fi, Type type, object originalValue)
    {
      if (fi.FieldType.Equals(Type.GetType("System.Guid")))
      {
        return originalValue.ToString();
      }
      else if (fi.FieldType.Equals(Type.GetType("System.Decimal")))
      {
        return originalValue.ToString();
      }
      else
        return base.WriteTransform(fi, type, originalValue);
    }
    
    public override object ReadTransform(FieldInfo fi, Type type, object serializeValue)
    {
      if (fi.FieldType.Equals(Type.GetType("System.Guid")))
      {
        Guid g = new Guid((string)serializeValue);
        return g;
      }
      else if (fi.FieldType.Equals(Type.GetType("System.Decimal")))
      {
        return Convert.ToDecimal((string)serializeValue);
      }
      else
        return base.ReadTransform(fi, type, serializeValue);
    }
    
    public override FieldType GetFieldType(FieldInfo fi, Type type)
    {
      if (fi.FieldType.Equals(Type.GetType("System.Guid")) || fi.FieldType.Equals(Type.GetType("System.Decimal")))
        return FieldType.STRING;
      return base.GetFieldType(fi, type);
    }
    
    public override bool IsIdentityField(FieldInfo fi, Type type)
    {
      if (fi.Name == "_identityField")
        return true;
      return base.IsIdentityField(fi, type);
    }
    public override string GetFieldName(FieldInfo fi, Type type)
    {
      if (fi.Name == "_nameChange")
        return fi.Name + "NewName";

      return fi.Name ;
    }

    public override bool IsFieldIncluded(FieldInfo fi, Type type)
    {
      if (fi.Name == "_notInclude")
        return false;
      return base.IsFieldIncluded(fi, type);
    }
  }
  
  //This demonstrates, how to map .NET type to pdx type or java type
  public class PdxTypeMapper : IPdxTypeMapper
  { 
    public string ToPdxTypeName(string localTypeName)
    {
      return "pdx_" + localTypeName;
    }

    public string FromPdxTypeName(string pdxTypeName)
    {
      return pdxTypeName.Substring(4);//need to extract "pdx_"
    }
  }
  
  // The PdxRemoteQuery QuickStart example.
  class PdxSerializer
  {
    static void Main(string[] args)
    {
      try
      {

        CacheFactory cacheFactory = CacheFactory.CreateCacheFactory();

        Console.WriteLine("Connected to the GemFire Distributed System");

        // Create a GemFire Cache with the "clientPdxRemoteQuery.xml" Cache XML file.
        Cache cache = cacheFactory.Set("cache-xml-file", "XMLs/clientPdxSerializer.xml").Create();

        Console.WriteLine("Created the GemFire Cache");

        // Get the example Region from the Cache which is declared in the Cache XML file.
        IRegion<string, Person> region = cache.GetRegion<string, Person>("Person");

        Console.WriteLine("Obtained the Region from the Cache");

        //to map .net type tp pdx type or java type
        Serializable.SetPdxTypeMapper(new PdxTypeMapper());
        
        // Register inbuilt reflection based autoserializer to serialize the domain types(Person class) as pdx format
        Serializable.RegisterPdxSerializer(new AutoSerializerEx());
        Console.WriteLine("Registered Person Query Objects");

        // Populate the Region with some PortfolioPdx objects.
        Person p1 = new Person("John", 1 /*ID*/, 23 /*age*/);
        Person p2 = new Person("Jack", 2 /*ID*/, 20 /*age*/);
        Person p3 = new Person("Tony", 3 /*ID*/, 35 /*age*/);
        
        region["Key1"] = p1;
        region["Key2"] = p2;
        region["Key3"] = p3;

        Console.WriteLine("Populated some Person Objects");

        //find the pool
        Pool pool = PoolManager.Find("examplePool");

        // Get the QueryService from the pool
        QueryService<string, Person> qrySvc = pool.GetQueryService<string, Person>();

        Console.WriteLine("Got the QueryService from the Pool");

        // Execute a Query which returns a ResultSet.    
        Query<Person> qry = qrySvc.NewQuery("SELECT DISTINCT * FROM /Person");
        ISelectResults<Person> results = qry.Execute();

        Console.WriteLine("ResultSet Query returned {0} rows", results.Size);

        // Execute a Query which returns a StructSet.
        QueryService<string, Struct> qrySvc1 = pool.GetQueryService<string, Struct>();
        Query<Struct> qry1 = qrySvc1.NewQuery("SELECT name, age FROM /Person WHERE id = 1");
        ISelectResults<Struct> results1 = qry1.Execute();

        Console.WriteLine("StructSet Query returned {0} rows", results1.Size);

        // Iterate through the rows of the query result.
        int rowCount = 0;
        foreach (Struct si in results1)
        {
          rowCount++;
          Console.WriteLine("Row {0} Column 1 is named {1}, value is {2}", rowCount, si.Set.GetFieldName(0), si[0].ToString());
          Console.WriteLine("Row {0} Column 2 is named {1}, value is {2}", rowCount, si.Set.GetFieldName(1), si[1].ToString());
        }

        
        // Close the GemFire Cache.
        cache.Close();

        Console.WriteLine("Closed the GemFire Cache");

      }
      // An exception should not occur
      catch (GemFireException gfex)
      {
        Console.WriteLine("PdxSerializer GemFire Exception: {0}", gfex.Message);
      }
    }
  }
}
